rust/src/delta.rs

```diff
@@ -613,6 +673,127 @@ impl std::fmt::Debug for DeltaTable {
+    fn from(error: StorageError) -> Self {
+        match error {
+            StorageError::AlreadyExists => DeltaTransactionError::VersionExists,
+            _ => DeltaTransactionError::Storage { source: error },
+        }
+    }
+}
+
+impl From<DeltaTableError> for DeltaTransactionError {
+    fn from(error: DeltaTableError) -> Self {
+        DeltaTransactionError::DeltaTable { source: error }
+    }
+}
+
+pub struct DeltaTransaction<'a> {
+    delta_table: &'a mut DeltaTable,
+}
+
+impl<'a> DeltaTransaction<'a> {
+    pub fn new(delta_table: &'a mut DeltaTable) -> Self {
+        DeltaTransaction { delta_table,}
+    }
+
+    pub async fn commit_all(&mut self, actions: &[Action], _operation: Option<DeltaOperation>) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        // TODO: stubbing `operation` parameter (which will be necessary for writing the CommitInfo action), but leaving it unused for now.
+        // `CommitInfo` is a fairly dynamic data structure so we should work out the data structure approach separately.
+
+        // TODO: calculate isolation level to use when checking for conflicts.
+        // Leaving conflict checking unimplemented for now to get the "single writer" implementation off the ground.
+        // Leaving some commmented code in place as a guidepost for the future.
+
+        // let no_data_changed = actions.iter().all(|a| match a {
+        //     Action::add(x) => !x.dataChange,
+        //     Action::remove(x) => !x.dataChange,
+        //     _ => false,
+        // });
+        // let isolation_level = if no_data_changed {
+        //     IsolationLevel::SnapshotIsolation
+        // } else { 
+        //     IsolationLevel::Serializable
+        // };
+
+        // TODO: create a CommitInfo action and prepend it to actions.
+
+        // try to commit in a loop in case other writers write the next version first
+        let version = self.try_commit_loop(actions).await?;
+
+        self.delta_table.update().await?;
+
+        Ok(version)
+    }
+
+    async fn try_commit_loop(&mut self, actions: &[Action]) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        let max_retries = max_retry_commit_attempts();
+        let mut attempt_number: u32 = 0;
+
+        loop {
+            if attempt_number > max_retries {
+                return Err(DeltaTransactionError::CommitRetriesExceeded);
+            }
+
+            let commit_result = self.try_commit(actions).await;
+
+            if commit_result.is_ok() {
+                return commit_result;
+            }
+
+            attempt_number += 1;
+        }
+    }
+
+    async fn try_commit(&mut self, actions: &[Action]) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        // Serialize all actions that are part of this log entry.
+        let jsons: Vec<String> = actions.iter()
+            .map(|a| {
+                serde_json::to_string(a).unwrap()
+            }).collect();
+        let log_entry = jsons.join("\n");
```

Comment:
looks like this is something that could be assembled by the caller instead? I can see a case where we might need to retry a commit with the same set of actions in append only mode due to version conflicts, in that case, we can avoid doing the same json serialization again.

also, we should propagate `serde_json::to_string` error to the caller :)

Reply:
fixed